### PR TITLE
Add encrypted tx and wrapper tx types

### PR DESCRIPTION
**Pull request by batconjurer**
This PR resolves the first two tasks in Issue #392. 

 * A WrapperTx type has been added that includes a encrypted tx payload and various metadata for verification.
 * This WrapperTx is included as data in a normal Tx type
 * During the CheckTx call from ABCI, a tx is forwarded to ProcessProposal. If the tx contains a wrapper tx as data, it is validated using the metadata. If validation fails, it is rejected from the mempool
 * Validations include deserializability, sufficient balance to pay fees, ciphertext validity checks from Ferveo, and signatures
 Non-encrypted Txs are still fully supported.

### CODE REVIEW

**Review by Acentelles — APPROVED** [review]

File: apps/src/lib/node/ledger/shell.rs [hunk 1]
```diff
@@ -495,16 +497,113 @@ impl Shell {
...
+                tx: req.tx,
+            };
+        }
+
+        let (result, processed_tx) = match process_tx(tx.unwrap()) {
+            // This occurs if the wrapper tx signature is invalid
+            Err(err) => (TxResult::from(err), None),
+            Ok(result) => match result {
+                // If it is a raw transaction, we do no further validation
+                TxType::Raw(_) => (
```
Comment (tzemanovic) [inline apps/src/lib/node/ledger/shell.rs:528]: this case is temporary until we encrypt all the transactions, right?

File: shared/Cargo.toml [hunk 2]
```diff
@@ -9,7 +9,15 @@ version = "0.1.0"
...
-default = []
+default = [
+  "tpke",
+  "ark-bls12-381",
+  "ark-ec",
+  "ark-serialize",
+  "rand_core",
+  "rand_new"]
+
+wasm = []
```
Comment (tzemanovic) [inline shared/Cargo.toml:20]: this is not used, is it?

File: shared/src/types/key/ed25519.rs [hunk 3]
```diff
@@ -448,6 +448,21 @@ impl From<PublicKey> for PublicKeyHash {
     }
 }
 
+impl From<&PublicKey> for PublicKeyHash {
```
Comment (tzemanovic) [inline shared/src/types/key/ed25519.rs:451]: we're adding this in #410

File: apps/src/lib/node/ledger/shell.rs (hunk 1 above)
Comment (batconjurer) [reply apps/src/lib/node/ledger/shell.rs:528]: Yes, in the sense that it is temporary. It allows backwards compatibility at the moment. In the future, we should also have `TxType::Decrypted(_)` and `TxType::UnDecryptable` and `TxType::PVSS` and maybe some others.

File: shared/Cargo.toml (hunk 2 above)
Comment (batconjurer) [reply shared/Cargo.toml:20]: Oops, that is a relic of an abandoned idea

File: shared/src/types/key/ed25519.rs (hunk 3 above)
Comment (batconjurer) [reply shared/src/types/key/ed25519.rs:451]: Then I will remove this when I merge in that PR.

**Review by tzemanovic — APPROVED** [review]

**Pull request merged**